Post: Cheat Sheet - Console Experience (#92)
PowerShell/Community-Blog: Posts/2022/07/cheat-sheet-console-experience.md

﻿---
post_title: Cheat Sheet - Console Experience

> username: Fred

Author: done

Categories: PowerShell
tags: PowerShell, Preference Variables, Convenience, Tab Expansion
Summary: How to have more control of preferences in functions and the role of modules on inheritance. 
---

# Cheat Sheet: Console Experience

PowerShell can take some getting used to.
Especially if you come at it from a different Shell and don't see any way to get your good old experience back.
However, hidden behind that plain white on blue shell, there is actually a wide range of customization options that help make your life less painful.
See below for the most commonly appreciated options.

## Tab Completion

The classic complaint we hear is that in Windows, Tab Completion is so much less helpful than for example in Bash.

> This is mostly because "CTRL" + "Space" is hard to discover, unless somebody shows you.

Reviewer: ```suggestion
This is mostly because <kdb>CTRL</kdb>+<kdb>Space</kdb> is hard to discover, unless somebody shows you.
```

## Key Bindings

There are a few keybindings that come in handy to know:

|CTRL + Space|Tab Menu|
|CTRL + r|Search in your input history|
|CTRL + a|Select everything in your current input/command line|
|CTRL + c|Copy everything currently selected in your input/command line to your clipboard|
|CTRL + v|Paste your clipboard into the current input/command line|
|SHIFT + Enter|Type multiline text in your console without executing the command|

Specifically, it is important to get used to not pasting with rightclick - by using CTRL+v instead, you get a single input history for multiple lines, you can preview your input before sending it (helps with those artifacts you get when pasting from Teams) and you stop accidentally overwriting your clipboard by selecting something in the console window.
​​​​​​​Also, with rightclick, you sometimes get the wrong order.

Oh, and you can [define your own keybindings](https://github.com/PowerShell/PSReadLine/blob/master/PSReadLine/SamplePSReadLineProfile.ps1) if you want to.
No need to accept the defaults.

## Packages

There are plenty of PowerShell packages out there that can make console life a lot less painful.

> Use Find-Module to search for them and Install-Module to install them. Example:

Reviewer: ```suggestion
Use `Find-Module` to search for them and `Install-Module` to install them. Example:
```

```powershell
Find-Module *SQL*
Install-Module Powerline
```

Looking for a command but don't know the module it is from?

```powershell
Find-Module -Command Write-PSFMessage
```

## Profile / Start Script

The key to ultimate customization is to have a way to define code that runs on each console start without requiring manual action.
Now if only there were a way to do that in PowerShell ...

```powershell
$profile​​​​​​​
```

Yeah, that simple. As long as that file exists, it will be run.

```powershell
notepad $profile
```

Add code, save, and you are good to go.

> There are different profile files per application running PowerShell - VSCode has a different one than pwsh.exe than powershell.exe.
> Make sure you edit the file you meant to edit.
> Or update the global profile for all applications: `$profile.CurrentUserAllHosts`

## PowerShell 7 / PowerShell Core

There's Windows PowerShell, which comes installed by default on any Windows.

> But there's also a [cool version you have to first install](https://github.com/powershell/powershell).

Reviewer: ```suggestion
But there's also a [cool version you have to first install](https://aka.ms/powershell-release?tag=stable).
```

It adds great convenience, better performance and the ability to actually like using Visual Studio Code with PowerShell.
You can grab it via a wide variety of sources, such as the Microsoft Store, Github or your preferred package manager.

You can also install it on MacOS or Linux.
​​​​​​​You should do so, it's awesome.

## Prompt

Want to customize your prompt to be more colorful / fancy / whatever else you want it to do?

> Well, all you need to do is override the function named "prompt" and put it in your profile and that's that.

Reviewer: ```suggestion
Well, all you need to do is override the function named `prompt` and put it in your profile and that's that.
```

Now if only there were commands in PowerShell to do so ...

```powershell
Get-Clipboard
Set-Clipboard
```

And since we're all about being lazy, there's aliases for that: "gcb" and "scb".
On that note, if you want to paste multiple columns into Excel, you want to use the tab delimiter.
Don't do manual labor though, here's the easy way to get data ready to paste to Excel:

```powershell
dir | ConvertTo-Csv -Delimiter "`t" | scb
```

That "ConvertTo-Csv" is way too much text though. Wouldn't it be nice to make that shorter?

## Aliases

In PowerShell, there is an easy way to be lazy: Aliases.
Use their power to abbreviate your commonly used commands.
Then put it in your $profile so you don't have to remember to add them.

```powershell
Set-Alias ctc ConvertTo-Csv
```

And now that previous line can be shortened to:

```powershell
dir | ctc -d "`t" | scb
```

Where is that "-d" coming from though?

## Short Parameter Names

You know all these commandline tools that have a long and a short notation for their parameters?
Like where you can either specify "--help" or "-h"?

Well, PowerShell takes that a step further:

> You only need to type a parameter so far until there's only one possibility what parameter it could be.

Reviewer: ```suggestion
You only need to type enough of the parameter name to uniquely identify it.
```

> Using the example above with "ConvertTo-Csv", there is only a single parameter that starts with "D", so that's enough to specify it.

Reviewer: ```suggestion
Using the example above with `ConvertTo-Csv`, there is only a single parameter that starts with "D", so that's enough to specify it.
```

> Actually, there is a common parameter named "Debug", but those don't count here.

## More Tools to improve the Console Experience

A lot more tools have been created to help being lazy & comfortable than could possibly all be listed, but here a few more projects out there that can help make console life more comfortable:

> |[AZ.Tools.Predictor](https://docs.microsoft.com/en-us/powershell/azure/az-predictor?view=azps-8.0.0)|Predictive intellisense for the AZ modules|

Reviewer: ```suggestion
|Tool|Description|
|---|---|
|[AZ.Tools.Predictor](https://docs.microsoft.com/powershell/azure/az-predictor)|Predictive intellisense for the AZ modules|
```

|[TabExpansionPlusPlus](https://www.powershellgallery.com/packages/TabExpansionPlusPlus/1.2)|Adds tab completion for classic commandline tools such as ROBOCOPY|
|[PSUtil](https://www.powershellgallery.com/packages/PSUtil)|Adds keybindings, aliases and other conveniences|
|[oh-my-posh](https://ohmyposh.dev/)|Transform your prompt, alternative to the Powerline module shown above|
|[posh-git](https://github.com/dahlbyk/posh-git)|Add git integration to your prompt|

## Paths, Explorer & PowerShell

Often enough you want to interact with the file system across applications:

+ Got the explorer open and want to start PowerShell in that path?
+ Just created an output file and want to open it?
+ Open the explorer in the current path?

For all of that there are convenient options.
From within the shell, "Invoke-Item" or its alias "ii" allow you to open a path in its default application:

```powershell
ii .\report.csv # Probably Excel
ii . # Current path in Explorer
```

The other way around works just as convenient:

> ​​​​​​​In the Windows Explorer, just type "pwsh.exe" (or "powershell.exe", if you didn't upgrade):

Reviewer: ```suggestion
​​​​​​​In the Windows Explorer, just type `pwsh.exe` (or `powershell.exe`, if you didn't upgrade):
```

![The Windows Explorer address bar, with the path replaced with "pwsh.exe"](./explorer.png)

## Inspecting Output

When you run a command, often enough you get some nice table, that is easy to read, but kind of lacking in data:

![Results of a dir command in a neat table](./format-table.png)

> Fortunately, by piping to "FL *", you get to see everything (even if it is a bit much):

Reviewer: ```suggestion
Fortunately, by piping to `FL *`, you get to see everything (even if it is a bit much):
```
